library/std/src/sys/unix/locks/futex_rwlock.rs

```diff
@@ -0,0 +1,293 @@
+use crate::sync::atomic::{
+    AtomicI32,
+    Ordering::{Acquire, Relaxed, Release},
+};
+use crate::sys::futex::{futex_wait, futex_wake, futex_wake_all};
+
+pub type MovableRwLock = RwLock;
+
+pub struct RwLock {
+    // The state consists of a 30-bit reader counter, a 'readers waiting' flag, and a 'writers waiting' flag.
+    // Bits 0..30:
+    //   0: Unlocked
+    //   1..=0x3FFF_FFFE: Locked by N readers
+    //   0x3FFF_FFFF: Write locked
+    // Bit 30: Readers are waiting on this futex.
+    // Bit 31: Writers are waiting on the writer_notify futex.
+    state: AtomicI32,
+    // The 'condition variable' to notify writers through.
+    // Incremented on every signal.
+    writer_notify: AtomicI32,
+}
+
+const READ_LOCKED: i32 = 1;
+const MASK: i32 = (1 << 30) - 1;
+const WRITE_LOCKED: i32 = MASK;
+const MAX_READERS: i32 = MASK - 1;
+const READERS_WAITING: i32 = 1 << 30;
+const WRITERS_WAITING: i32 = 1 << 31;
+
+fn unlocked(state: i32) -> bool {
+    state & MASK == 0
+}
+
+fn write_locked(state: i32) -> bool {
+    state & MASK == WRITE_LOCKED
+}
+
+fn readers_waiting(state: i32) -> bool {
+    state & READERS_WAITING != 0
+}
+
+fn writers_waiting(state: i32) -> bool {
+    state & WRITERS_WAITING != 0
+}
+
+fn read_lockable(state: i32) -> bool {
+    // This also returns false if the counter could overflow if we tried to read lock it.
+    state & MASK < MAX_READERS && !readers_waiting(state) && !writers_waiting(state)
+}
+
+fn reached_max_readers(state: i32) -> bool {
+    state & MASK == MAX_READERS
+}
+
+impl RwLock {
+    #[inline]
+    pub const fn new() -> Self {
+        Self { state: AtomicI32::new(0), writer_notify: AtomicI32::new(0) }
+    }
+
+    #[inline]
+    pub unsafe fn destroy(&self) {}
+
+    #[inline]
+    pub unsafe fn try_read(&self) -> bool {
+        self.state
+            .fetch_update(Acquire, Relaxed, |s| read_lockable(s).then(|| s + READ_LOCKED))
+            .is_ok()
+    }
+
+    #[inline]
+    pub unsafe fn read(&self) {
+        if !self.try_read() {
+            self.read_contended();
+        }
+    }
+
+    #[inline]
+    pub unsafe fn read_unlock(&self) {
+        let state = self.state.fetch_sub(READ_LOCKED, Release) - READ_LOCKED;
+
+        // It's impossible for a reader to be waiting on a read-locked RwLock,
+        // except if there is also a writer waiting.
+        debug_assert!(!readers_waiting(state) || writers_waiting(state));
+
+        // Wake up a writer if we were the last reader and there's a writer waiting.
+        if unlocked(state) && writers_waiting(state) {
+            self.wake_writer_or_readers(state);
+        }
+    }
+
+    #[cold]
+    fn read_contended(&self) {
+        let mut state = self.spin_read();
+
+        loop {
+            // If we can lock it, lock it.
+            if read_lockable(state) {
+                match self.state.compare_exchange(state, state + READ_LOCKED, Acquire, Relaxed) {
```

Comment:
Sure. I don't think it really matters though. Have you ever seen llvm generate anything more efficient when switching from compare_exchange to compare_exchange_weak in a loop? It doesn't actually turn a loop with compare_exchange_weak into a basic ll/sc loop, and still uses a regular load at the start of the loop and only uses the load-link (ldaxr) at the end of the loop with a separate compare and branch-not-equal before the store-conditional. (Because otherwise it can't guarantee that the linked load and store are close enough to each other to guarantee progress.)

Reply:
weak allows it to run loop conditions more frequently (on stxr failure instead of on cmp failure/clrex). But that shouldn't make a difference here as the condition would be the same in the strong case.

Reply:
On platforms using LL/SC you will end up with 2 nested loops: the outer loop + an inner loop for the strong compare_exchange. You can eliminate the inner loop by using compare_exchange_weak. This isn't too significant here since this is the slow path, but it is significant in the inline fast path.